crates/topos-tce-api/src/runtime/sync_task.rs

```diff
@@ -0,0 +1,179 @@
+    /// The connection to the database layer through a StorageClient
+    pub(crate) storage: StorageClient,
+    /// The notifier is used to send certificates to the stream
+    pub(crate) notifier: Sender<StreamCommand>,
+    /// If a new stream is registered with the same Uuid, the sync task will be cancelled
+    pub(crate) cancel_token: CancellationToken,
+}
+
+impl SyncTask {
+    /// Creating a new SyncTask which will fetch certificates from the storage and pushes them to the stream
+    pub(crate) fn new(
+        stream_id: Uuid,
+        target_subnet_stream_positions: TargetSubnetStreamPositions,
+        storage: StorageClient,
+        notifier: Sender<StreamCommand>,
+        cancel_token: CancellationToken,
+    ) -> Self {
+        println!("CREATED A NEW TASK");
+        Self {
+            status: SyncTaskStatus::Running,
+            stream_id,
+            target_subnet_stream_positions,
+            storage,
+            notifier,
+            cancel_token,
+        }
+    }
+}
+
+impl IntoFuture for SyncTask {
+    type Output = SyncTaskStatus;
+
+    type IntoFuture = Pin<Box<dyn Future<Output = Self::Output> + Send + 'static>>;
+
+    fn into_future(mut self) -> Self::IntoFuture {
+        Box::pin(async move {
+            println!("START FUTURE");
+            info!("Sync task started for stream {}", self.stream_id);
+            let mut collector: Vec<(CertificateDelivered, FetchCertificatesPosition)> = Vec::new();
+
+            for (target_subnet_id, source) in &mut self.target_subnet_stream_positions {
+                if self.cancel_token.is_cancelled() {
+                    self.status = SyncTaskStatus::Cancelled;
+                    return SyncTaskStatus::Cancelled;
+                }
+                let source_subnet_list = self
+                    .storage
+                    .get_target_source_subnet_list(*target_subnet_id)
+                    .await;
+
+                info!(
+                    "Stream sync task detected {:?} as source list",
+                    source_subnet_list
+                );
+                if let Ok(source_subnet_list) = source_subnet_list {
+                    for source_subnet_id in source_subnet_list {
+                        if let Entry::Vacant(entry) = source.entry(source_subnet_id) {
+                            entry.insert(TargetStreamPosition {
+                                target_subnet_id: *target_subnet_id,
+                                source_subnet_id,
+                                position: 0,
+                                certificate_id: None,
+                            });
+                        }
+                    }
+                }
+
+                for TargetStreamPosition {
+                    target_subnet_id,
+                    source_subnet_id,
+                    position,
+                    ..
+                } in source.values_mut()
+                {
+                    if self.cancel_token.is_cancelled() {
```

Comment:
Why check this again? After line 85 we know the task is not cancelled yeah?

Reply:
There is an `async` call happening in each loop, so I wonder, what if this call takes longer and we want to cancel in between as well?